enhancements/cel-scanner.md

```diff
@@ -0,0 +1,508 @@
+  scan := &ComplianceScan{}
+  // ...
+  if tp.GetAnnotations()["compliance.openshift.io/scanner-type"] == "cel" {
+    scan.Spec.ScannerType = compv1alpha1.ScannerTypeCelScanner
+  } else {
+    scan.Spec.ScannerType = compv1alpha1.ScannerTypeOpenSCAP
+  }
+  return scan
+}
+```
+
+### 6. TailoredProfile Controller
+
+- Ensures that if a `TailoredProfile` references a `CustomRule` (with `scannerType: cel`), the resulting annotation `scanner-type: cel` is set.  
+- For **custom** TailoredProfiles (where you only have `CustomRule`s and no `ProfileBundle` or “extends” parent), the operator will skip the SCAP `tailoredProfile` Datastream `ConfigMap` generation. Instead, it sets `annotation[scanner-type]=cel`.
+
+**Pseudo-code**:
+
+```go
+func (r *ReconcileTailoredProfile) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+  tp := &TailoredProfile{}
+  err := r.client.Get(ctx, request.NamespacedName, tp)
+  // ...
+  foundCel := false
+  for each ruleRef in tp.Spec.EnableRules {
+     if isCustomRule(ruleRef) {
+       foundCel = true
+       break
+     }
+  }
+  if foundCel {
+     // ensure annotation is set
+     ann := tp.GetAnnotations()
+     ann["compliance.openshift.io/scanner-type"] = "cel"
+     tp.SetAnnotations(ann)
+     r.client.Update(ctx, tp)
+  }
+  // also make sure only one type of rule are there
+  // ...
+}
+```
+
+### User Stories
+
+1. **On the fly Custom Rule ComplianceCheck**
+   An admin wants to confirm each non-control-plane namespace has at least one `NetworkPolicy`. They define a `CustomRule` referencing `namespaces` and `networkpolicies`.  
+   - The “cel-scanner” container fetches these resources, checks the CEL expression, and evaulate to a pass/fail turns this into a `ComplianceCheckResult`.
+
+### API Extensions
+
+- **`CustomRule` CRD**:  
+  - A new CRD Extends existing `ComplianceRule` logic with CEL-specific fields.  
+
+- **`TailoredProfile` CRD**:  
+  - Add Type filed in Spec.enableRules 
+  - Extended to reference `CustomRule` objects under `enableRules[]`.  
+  - Sets a `scannerType` annotation if only `CustomRule`s are enabled.
+
+- **`ComplianceScan` CRD**:  
+  - Add new field `scannerType` in Spec
+
+### Implementation Details / Notes / Constraints
+
+- **RBAC**: The “cel-scanner” Pod needs read permissions (`get/list/watch`) for every resource type mentioned in the `inputs`.
+- **Performance**: Large-scale resource fetching can be costly if a `CustomRule` references a wide resource scope.
+
+### Risks and Mitigations
+
+- **Excessive Resource Fetching**: Large queries can slow the cluster. Mitigate by documenting best practices (narrow `namespace`, etc.).  
+- **Security**: Broad `get/list/watch` for multiple resource types. Admins can carefully scope or block unneeded RBAC for `CustomRule`s.
+
+## Design Details
+
+
+### The Implementation “Flow”
+
+Putting it all together:
+
+1. **User** defines a `CustomRule` with `scannerType:"cel"`.
```

Comment:
Similar to the above question.

If I write a `CustomRule` to evaluate a resource that isn't include in the API resources aggregator permissions, do I need a subsequent step here to make sure it can fetch those resources?

Otherwise, if I understand correctly, it's entirely possible that a user would get all the way through the flow, up to the scan, only to have it fail on a `403`?

Reply:
I think that's a good point, yeah, we can make it that we check for the permission before a scan is happened.